src/util/threadpool.rs

```diff
@@ -82,21 +82,118 @@ pub trait ScheduleQueue<T: Debug> {
-        GroupStatisticsItem {
-            running_count: 0,
-            high_priority_queue_count: 0,
+    fn sum(&self) -> usize {
+        self.running_count + self.high_priority_queue_count + self.low_priority_queue_count
+    }
+}
+
+// `SmallGroupFirstQueue` tries to run groups with number of
+//  tasks smaller than small_group_tasks_limit first when all threads are busy.
+pub struct SmallGroupFirstQueue<T> {
+    high_priority_queue: VecDeque<Task<T>>,
+    low_priority_queue: VecDeque<Task<T>>,
+    group_concurrency_on_busy: usize,
+    small_group_tasks_limit: usize,
+    statistics: HashMap<T, GroupStatisticsItem>,
+}
+
+impl<T: Hash + Eq + Ord + Send + Clone + Debug> SmallGroupFirstQueue<T> {
+    pub fn new(group_concurrency_on_busy: usize,
+               small_group_tasks_limit: usize)
+               -> SmallGroupFirstQueue<T> {
+        SmallGroupFirstQueue {
+            high_priority_queue: VecDeque::new(),
+            low_priority_queue: VecDeque::new(),
+            statistics: HashMap::new(),
+            group_concurrency_on_busy: group_concurrency_on_busy,
+            small_group_tasks_limit: small_group_tasks_limit,
         }
     }
 
-    fn sum(&self) -> usize {
-        self.running_count + self.high_priority_queue_count
+    fn pop_from_high_priority_queue(&mut self) -> Option<Task<T>> {
+        if self.high_priority_queue.is_empty() {
+            return None;
+        }
+        let task = self.high_priority_queue.pop_front().unwrap();
+        let mut statistics = self.statistics.get_mut(&task.gid).unwrap();
+        statistics.high_priority_queue_count -= 1;
+        Some(task)
+    }
+
+    fn pop_from_low_priority_queue(&mut self) -> Option<Task<T>> {
+        if self.low_priority_queue.is_empty() {
+            return None;
+        }
+        let task = self.low_priority_queue.pop_front().unwrap();
+        let mut statistics = self.statistics
+            .get_mut(&task.gid)
+            .unwrap();
+        statistics.low_priority_queue_count -= 1;
+        Some(task)
+    }
+}
+
+impl<T: Hash + Eq + Ord + Send + Clone + Debug> ScheduleQueue<T> for SmallGroupFirstQueue<T> {
+    fn push(&mut self, task: Task<T>) {
+        let mut statistics = self.statistics
+            .entry(task.gid.clone())
+            .or_insert(Default::default());
+        if statistics.low_priority_queue_count == 0 &&
+           statistics.running_count + statistics.high_priority_queue_count <
+           self.small_group_tasks_limit {
+            self.high_priority_queue.push_back(task);
+            statistics.high_priority_queue_count += 1;
+        } else {
+            self.low_priority_queue.push_back(task);
+            statistics.low_priority_queue_count += 1;
+        }
+    }
+
+    fn pop(&mut self) -> Option<Task<T>> {
+        if self.low_priority_queue.is_empty() {
+            return self.pop_from_high_priority_queue();
+        }
+
+        if self.high_priority_queue.is_empty() {
```

Comment:
IMO, it's easier to understand to check the high_priority_queue first since it's high priority. so the codes look better:

```
if self.high_priority_queue.is_empty() {
 ...
}

if self.low_priority_queue.is_empty() {
  ...
}
```